atomisticparsers/gromacs/parser.py

```diff
@@ -1250,8 +1406,80 @@ def parse_workflow(self):
             else:
                 method["thermodynamic_ensemble"] = "NVE"
 
+            params_key = "free_energy_calculation_parameters"
+            method[params_key] = self.get_free_energy_calculation_parameters()
+
+            self.xvg_parser.mainfile = self.get_gromacs_file("xvg")
+            free_energies = self.xvg_parser.get("results")
+
+            title = free_energies.get("title", "") if free_energies is not None else ""
+            flag_FE = False
+            if r"dH/d\xl\f{}" in title and r"\xD\f{}H" in title:
+                flag_FE = True
+                results_key = "free_energy_calculations"
+                results[results_key] = {}
+                columns = free_energies.get("column_vals")
+                results[results_key]["n_frames"] = len(columns)
+                results[results_key]["n_states"] = len(
+                    method[params_key].get("lambdas", [])
+                )
+                results[results_key]["lambda_index"] = method[params_key].get(
+                    "lambda_index", None
+                )
+                results[results_key]["times"] = columns[:, 0] * ureg.ps
+                columns = columns[:, 1:] * self._gro_energy_units.magnitude
+                results[results_key]["value_unit"] = str(self._gro_energy_units.units)
+
             self.parse_md_workflow(dict(method=method, results=results))
 
+            if flag_FE:
+                filename = os.path.join(
+                    os.path.dirname(self.filepath.split("/raw/")[-1]),
+                    f"{os.path.basename(self.filepath)}.archive.hdf5",
+                )
+                if not os.path.isfile(
+                    os.path.join(os.path.dirname(self.filepath), filename)
+                ):
+                    if self.archive.m_context:
+                        with self.archive.m_context.raw_file(filename, "wb") as f:
+                            pass
+                sec_FE_parameters = (
+                    self.archive.workflow2.method.free_energy_calculation_parameters[0]
+                )
+                sec_FE = self.archive.workflow2.results.free_energy_calculations[0]
+                sec_FE.method_ref = sec_FE_parameters
+                if self.archive.m_context:
+                    with self.archive.m_context.raw_file(filename, "r+b") as f:
+                        # The expected columns of the xvg file are:
+                        # Total Energy
+                        # dH/dlambda current lambda
+                        # Delta H between each lambda and current lambda (n_lambda columns)
+                        # PV Energy
+                        if columns[:, 2:-1].shape[1] != sec_FE.n_states:
+                            self.logger.warning(
+                                "Unexpected format of xvg file. Not storing free energy calculation results."
+                            )
```

Comment:
are we still supposed to write if check fails?

Reply:
no, you are right, I have moved this check above and added a couple others.